[#8] 게시글 도메인 구현

JIWON27 opened the pull request:
* 게시글 도메인 구현
* 게시글 CRUD 구현

JIWON27 committed:
feat: 챌린지 ID를 가지고 IN절 조회하는 기능 추가

JIWON27 committed:
feat: 유저가 참여한 챌린지 조회/존재하는지 확인하는 기능 추가

JIWON27 committed:
style: 불필요한 공백 삭제

JIWON27 committed:
feat: 요약된 챌린지 리스트 응답 fixture 추가

JIWON27 committed:
feat: 예외 메시지를 받도록 추가

JIWON27 committed:
feat: UserResponse 반환값 Detail, Summary 추가
- UserResponse 반환값 Detail, Summary 추가에 따른 Serivice, Controller 코드 수정
- Service, Controller 수정에 따른 테스트 코드 fixtrue 추가 및 테스트 코드 수정

JIWON27 committed:
chore: QueryDSL 의존성 추가

JIWON27 committed:
feat: Post, PostChallenge 엔티티 추가

JIWON27 committed:
feat: Post 엔티티 관련 enums, repository 구현
- 게시글 카테고리, 정렬 enum 추가
- QueryDSL을 사용한 게시글 조회 시 필터링 기능 추가

JIWON27 committed:
feat: QueryDSL 설정 추가

JIWON27 committed:
feat: 게시글 Service, Controller 구현

JIWON27 committed:
feat: 컴포넌트 스캔 범위 지정

JIWON27 committed:
feat: 게시글 HTTP 요청 파일

rnignon commented on module-api/src/main/java/com/trybe/moduleapi/post/controller/PostController.java:
```diff
@@ -0,0 +1,54 @@
+        return postService.save(userDetails.getUser(), request);
+    }
+
+    @GetMapping("/{id}")
+    public PostResponse.Detail findById(@PathVariable Long id) {
+        return postService.find(id);
+    }
+
+    // 전체 조회 (정렬/필터링이 없는 경우 + 정렬/필터링이 있는 경우)
+    @PostMapping("/search")
+    public PageResponse<PostResponse.Summary> findAll(@Valid @RequestBody PostRequest.Read request,
+                                                     Pageable pageable){
+        return postService.findAll(request, pageable);
+    }
```
SecurityConfig에 해당 엔드포인트 추가 하셨나요??

JIWON27 replied:
저는 커뮤니티쪽은 로그인 후 사용할 수 있도록 할려고했는데 조회랑 전체조회까지는 풀어두는것도 좋겠네요~

rnignon commented on module-api/src/main/java/com/trybe/moduleapi/post/dto/PostRequest.java:
```diff
@@ -0,0 +1,75 @@
+    private static final String TITLE_VALIDATION_MESSAGE = "제목을 입력해주세요.";
+    private static final int TITLE_MESSAGE_MIN = 10;
+    private static final int TITLE_MESSAGE_MAX = 20;
+    private static final String TITLE_MESSAGE_SIZE_VALIDATION_MESSAGE = "내용은 10자~20자 사이로 입력해주세요.";
```
메시지가 제목이 아니라 내용으로 되어있어요!
그리고 최대 20자는 조금 타이트할 것 같아요...,,

rnignon commented on module-api/src/main/java/com/trybe/moduleapi/post/dto/PostRequest.java:
```diff
@@ -0,0 +1,75 @@
+import java.util.Set;
+
+public class PostRequest {
+    private static final String CATEGORIES_NOT_EMPTY_MESSAGE = "포스트 카테고리를 선택해주세요.";
```
게시글이나 포스트 중 하나로 통일해서 사용하는 게 좋을 것 같아요!

rnignon commented on module-api/src/main/java/com/trybe/moduleapi/post/dto/PostResponse.java:
```diff
@@ -0,0 +1,50 @@
+                              post.getTitle(),
+                              post.getContent(),
+                              post.getCategory(),
+                              com.trybe.moduleapi.user.dto.response.UserResponse.Summary.from(post.getUser()),
```
이거 패키지 경로 제거해주실 수 있나요?

rnignon commented on module-api/src/main/java/com/trybe/moduleapi/post/dto/PostResponse.java:
```diff
@@ -0,0 +1,50 @@
+            PostCategory category,
+            UserResponse.Summary user,
+            LocalDateTime createdAt,
+            List<ChallengeResponse.Summary> challengeSummary
```
혹시 이 필드 이름 challenges로 변경해주실 수 있을까요? 프론트 측에서 조금 더 직관적으로 인식할 수 있을 것 같아요

rnignon commented on module-api/src/main/java/com/trybe/moduleapi/post/dto/PostResponse.java:
```diff
@@ -0,0 +1,50 @@
+            String content,
+            PostCategory category,
+            UserResponse.Summary user,
+            @JsonFormat(pattern = "yyyy-MM-dd'T'HH:mm:ss") // JSON 직렬화 시 포맷 적용
```
이거 사용하신다면 Detail에도 동일하게 넣어주면 좋을 것 같아요~

rnignon commented on module-api/src/main/java/com/trybe/moduleapi/post/dto/PostResponse.java:
```diff
@@ -0,0 +1,50 @@
+    public record Summary(
+            Long id,
+            String title,
+            String content,
```
Summary 응답에는 content가 불필요할 것 같아요!

rnignon commented on module-api/src/main/java/com/trybe/moduleapi/post/dto/PostRequest.java:
```diff
@@ -0,0 +1,75 @@
+            @NotNull(message = CATEGORIES_NOT_EMPTY_MESSAGE)
+            PostCategory category,
+
+            Set<Long> challengeId
```
챌린지 ID가 여러 개 들어있는 Set을 표현하는 것이니 필드명을 challengeIds로 변경하는 건 어떠신가요?

rnignon commented on module-api/src/main/java/com/trybe/moduleapi/post/dto/PostResponse.java:
```diff
@@ -0,0 +1,50 @@
+                              post.getCategory(),
+                              com.trybe.moduleapi.user.dto.response.UserResponse.Summary.from(post.getUser()),
+                              post.getCreatedAt(),
+                              summaryChallenge);
```
이 부분도 User 처리한 것처럼 객체 자체를 받은 후에 Response로 여기서 변환해주는 게 어떨까요?

rnignon commented on module-api/src/main/java/com/trybe/moduleapi/post/service/PostService.java:
```diff
@@ -0,0 +1,142 @@
+
+        List<ChallengeResponse.Summary> summaryChallenges = savePostChallenge(post, request.challengeId());
+        return PostResponse.Detail.from(savePost ,summaryChallenges);
+    }
```
해당 메서드에서 challengeId가 비어있을 때, savePostChallenge 메서드가 수행되면 빈 리스트가 반환되지 않나요??
만약 빈 리스트에 대해서 정상적으로 수행된다면, isEmpty 파트가 생략되어도 될 것 같아요

rnignon commented on module-api/src/main/java/com/trybe/moduleapi/post/service/PostService.java:
```diff
@@ -0,0 +1,142 @@
+    @Transactional(readOnly = true)
+    public PageResponse<PostResponse.Summary> findAll(PostRequest.Read request, Pageable pageable){
+        Page<Post> posts = postRepository.findAllByKeywordAndCategories(request.keyword(), request.categories(), request.order(), pageable);
+        Page<PostResponse.Summary> summaryPostPageDate = posts.map(PostResponse.Summary::from);
```
summaryPostPageDate 중 Date는 무슨 뜻인가요??

JIWON27 replied:
그러게요.. 수정하겠습니다😅

rnignon commented on module-api/src/main/java/com/trybe/moduleapi/post/service/PostService.java:
```diff
@@ -0,0 +1,142 @@
+
+        checkLoginUserAndPostUser(user, post);
+
+        postRepository.deleteById(id);
+        postChallengeRepository.deleteAllByPostId(post.getId());
```
여기 삭제 부분에서 PostChallenge를 먼저 삭제한 후 Post를 삭제해야 참조 무결성 오류가 발생하지 않을 것 같아요

JIWON27 replied:
현재 SoftDelete 상태여서 참조 무결성 문제가 발생은 안했습니다~
근데 말씀해주신 것처럼 안정성을 위해서 수정하겠습니다~

rnignon commented on module-core/src/main/java/com/trybe/modulecore/post/enums/PostOrder.java:
```diff
@@ -0,0 +1,25 @@
+    LATEST("최신순") {
+        @Override
+        public OrderSpecifier<?> getOrderSpecifier() {
+            return post.id.desc();
```
createdAt 필드를 사용하는 것은 어떨까요?? 더 직관적이고, ID 명명 규칙이 변경되더라도 잘 동작할 수 있을 것 같아요

rnignon commented on module-core/src/main/java/com/trybe/modulecore/post/repository/PostCustomRepositoryImpl.java:
```diff
@@ -0,0 +1,64 @@
+        List<Post> result = jpaQueryFactory.select(post)
+                                           .from(post)
+                                           .where(containsKeyword(keyword), filterCategories(categories))
+                                           .offset(pageable.getPageNumber() * pageable.getPageSize())
```
pageable의 offset을 사용해도 좋을 것 같아요~

rnignon reviewed (changes requested):
수고하셨습니다요~~ 🫡

JIWON27 committed:
feat: POST 조회 API 추가

JIWON27 committed:
feat: POST DTO 수정
- PostRequest 게시글 제목 길이제한 50으로 변경
- challengId를 challengIds로 수정

JIWON27 committed:
refactor: PostService 변경
- POST DTO 변경에 따른 서비스 코드 수정
- challengeIds 비어있는지 체크하는 로직 삭제
- savePostChallenge(...) 메서드 분리

JIWON27 committed:
refactor: 최신순 createAt으로 변경

JIWON27 committed:
refactor: Pageable의 getOffset을 사용하도록 변경

JIWON27 committed:
fix: POST 수정 API 엔드포인트 누락 해결
